File: src/main/java/org/beatonma/lib/ui/pref/PreferenceLoader.java
package org.beatonma.lib.ui.pref;

import android.content.Context;

import org.beatonma.lib.load.AsyncResult;
import org.beatonma.lib.load.SupportBaseAsyncTaskLoader;
import org.beatonma.lib.ui.pref.preferences.PreferenceGroup;

/**
 * Asynchronously parse a preferences JSON file and return a PreferenceGroup instance
 */
public class PreferenceLoader extends SupportBaseAsyncTaskLoader<PreferenceGroup> {
    private final int mResourceID;

    public PreferenceLoader(final Context context, final int resourceID) {
        super(context);
        mResourceID = resourceID;
    }

    @Override
    public AsyncResult<PreferenceGroup> loadInBackground() {
        final AsyncResult.Builder<PreferenceGroup> result = AsyncResult.getBuilder();
        final PreferenceGroup prefs = PreferenceGroup.fromJson(getContext(), mResourceID);

        if (prefs.isEmpty()) {
            result.failure("Loaded preferences are empty - please check your definitions!");
            return result;
        }

        prefs.load(getContext());
        result.success(prefs);
        return result;
    }

    @Override
    protected void onReleaseResources(final AsyncResult data) {

    }
}

File: src/main/java/org/beatonma/lib/ui/pref/core/PreferenceLoader.kt
package org.beatonma.lib.ui.pref.core

import android.content.Context
import org.beatonma.lib.load.Result
import org.beatonma.lib.load.SupportBaseAsyncTaskLoader
import org.beatonma.lib.ui.pref.preferences.PreferenceGroup
import org.beatonma.lib.ui.pref.preferences.buildPreferencesFromJson

/**
 * Asynchronously parse a preferences JSON file and return a PreferenceGroup instance
 */
class PreferenceLoader(
        context: Context,
        private val mResourceID: Int
) : SupportBaseAsyncTaskLoader<PreferenceGroup>(context) {

    override fun loadInBackground(): Result<PreferenceGroup> {
        val result = Result.getBuilder<PreferenceGroup>()
        val prefs: PreferenceGroup
        try {
            prefs = buildPreferencesFromJson(context, mResourceID)
        } catch (e: Exception) {
            result.failure(e)
            return result
        }

        if (prefs.isEmpty) {
            result.failure("Loaded preferences are empty - please check your definitions!")
            return result
        }

        result.success(prefs)
        return result
    }

    override fun onReleaseResources(data: Result<PreferenceGroup>?) {

    }
}
